src/main/kotlin/org/malachite/estella/services/ApplicationService.kt

```diff
@@ -55,27 +57,53 @@ class ApplicationService(
         val recruitmentProcessStages = recruitmentProcess
             .stages
             .sortedBy { it.id }
-        val index = recruitmentProcessStages.indexOf(application.stage)
-        val newStage = recruitmentProcessStages[index + 1]
-        if (index == recruitmentProcessStages.lastIndex - 1)
-            applicationRepository.save(
-                application.copy(
-                    stage = newStage,
-                    status = ApplicationStatus.ACCEPTED
-                )
-            )
-        else if (index < recruitmentProcessStages.lastIndex)
-            applicationRepository.save(application.copy(stage = recruitmentProcessStages[index + 1]))
-
-        when (newStage.type) {
-            StageType.HR_INTERVIEW, StageType.TECHNICAL_INTERVIEW -> interviewService.createInterview(applicationId, newStage.type)
-            else -> null
+
+        val applicationRecruitmentStages = application.applicationStages.map { it.stage }.sortedBy { it.id }
+
+        if (applicationRecruitmentStages.isEmpty()) {
+            application.addNewApplicationStageData(recruitmentProcess.stages.first())
+            return
         }
+
+        val indexOfRecruitmentStage = recruitmentProcessStages.indexOf(applicationRecruitmentStages.last())
+
+        if (isNotLastStage(indexOfRecruitmentStage, recruitmentProcessStages.lastIndex))
+            application.addNewApplicationStageData(recruitmentProcessStages[indexOfRecruitmentStage + 1])
+                .let {
+                    if (shouldBeAccepted(indexOfRecruitmentStage, recruitmentProcessStages.lastIndex))
+                        it.copy(status = ApplicationStatus.ACCEPTED)
+                    else
+                        it
+                }.let {
+                        applicationRepository.save(it)
+                        val newStage = recruitmentProcessStages[indexOfRecruitmentStage]
+                        it.applicationStages.find { it.stage == newStage }?.let {
+                            when (newStage.type) {
+                                StageType.HR_INTERVIEW, StageType.TECHNICAL_INTERVIEW ->
+                                    interviewService.createInterview(it, InterviewPayload(null, null))
+                                else -> null
+                            }
+                        }
+                        it
+                }
```

Comment:
Czemu tego nie robimy w ApplicationStageDataService.createApplicationStageData? (Czy to się nam nie duplikuje z tym co tam jest)

Reply:
Ty tam tworzysz nowe Interview?

Reply:
masz racje, jest duplikacja z tym co tam robisz